### PR TITLE
emulator for Phase-2 Calo PF clusters using GCT firmware

### PR DESCRIPTION
**Pull request by pallabidas**
#### PR description:

  - adds a data format: l1tp2::CaloPFClusterCollection 
  - depends on PR [Devel phase2eg standalone barrel emulator #1069](https://github.com/cms-l1t-offline/cmssw/pull/1069) for input GCT towers collection (ECAL clusters + HCAL towers)
  - refer to presentation here: https://indico.cern.ch/event/1209727/#2-pf-clusters-from-gct

#### PR validation:

has passed code checks in draft PR [#1075](https://github.com/cms-l1t-offline/cmssw/pull/1075) and validation plots look ok for the collections created.

resolution_genPt_barrel

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

not a backport

### CODE REVIEW

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`cmsenv && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`cmsenv && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`cmsenv && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no files with code format issues!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`cmsenv && scram b -k -j 8 code-format-all`|

File: L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  if (!iEvent.getByToken(caloTowerToken_, caloTowerCollection))
+    edm::LogError("Phase2L1CaloPFClusterEmulator") << "Failed to get towers from caloTowerCollection!";
+
+  iEvent.getByToken(caloTowerToken_, caloTowerCollection);
+  float GCTintTowers[nTowerEta][nTowerPhi];
+  float realEta[nTowerEta][nTowerPhi];
+  float realPhi[nTowerEta][nTowerPhi];
+  for (const l1tp2::CaloTower& i : *caloTowerCollection) {
+    //std::cout << "Phase2L1CaloPFClusterEmulator: GCT FULL tower energy " << i.ecalTowerEt()
+    //          << " at eta, phi: (" << i.towerIEta() << ", " << i.towerIPhi() << ")" << std::endl;
```
Comment (aloeliger) [inline L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc]: Please remove commented code

File: L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    tempPfclusters = gctpf::pfcluster(temporary, etaoffset, phioffset);
+
+    for (int i = 0; i < nPFClusterSLR; i++) {
+      int gcteta = tempPfclusters.GCTpfclusters[i].eta;
+      int gctphi = tempPfclusters.GCTpfclusters[i].phi;
+      float towereta = realEta[gcteta][gctphi];
+      float towerphi = realPhi[gcteta][gctphi];
+      //std::cout << "Phase2L1CaloPFClusterEmulator: GCT Pfclusters energy " << tempPfclusters.GCTpfclusters[i].et
+      //          << " at ieta, iphi: (" << tempPfclusters.GCTpfclusters[i].eta << ", " << tempPfclusters.GCTpfclusters[i].phi << ")"
+      //          << " at eta, phi: (" << towereta << ", " << towerphi << ")" << std::endl;
```
Comment (aloeliger) [inline L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc]: Please remove commented code

**Comment by pallabidas** [comment]
hi @aloeliger the branch is updated to work on the latest checkout instructions for phase2-l1t-integration-1252patch1 , please check

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 2 files that did not meet formatting requirements:
 - L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc
 - L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

File: L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void Phase2L1CaloPFClusterEmulator::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
+}
+
```
Comment (aloeliger) [inline L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc]: They're getting picky about this, so this module should have a description.

File: L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+// ------------ method called to produce the data  ------------
+void Phase2L1CaloPFClusterEmulator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  std::unique_ptr<l1tp2::CaloPFClusterCollection> pfclusterCands(make_unique<l1tp2::CaloPFClusterCollection>());
+
+  edm::Handle<std::vector<l1tp2::CaloTower>> caloTowerCollection;
+  if (!iEvent.getByToken(caloTowerToken_, caloTowerCollection))
+    edm::LogError("Phase2L1CaloPFClusterEmulator") << "Failed to get towers from caloTowerCollection!";
+
+  iEvent.getByToken(caloTowerToken_, caloTowerCollection);
```
Comment (aloeliger) [inline L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc]: You check if you can get this, and log an error... but just proceed onward to try and get this anyways? Is this module tolerant to this error? Should this throw a `cms::exception` instead if these are not available?

File: DataFormats/L1TCalorimeterPhase2/src/classes_def.xml [hunk 5]
```diff
@@ -42,5 +42,13 @@
   </class>
   <class name="l1tp2::DigitizedClusterGTCollection" />
   <class name="edm::Wrapper<l1tp2::DigitizedClusterGTCollection>" /> 
+
+  <class name="l1tp2::CaloPFCluster"  ClassVersion="4">
+         <version ClassVersion="4" checksum="4176166177"/>
+  </class>
+  <class name="std::vector<l1tp2::CaloPFCluster>" />
+  <class name="l1tp2::CaloPFClusterCollection" />
+  <class name="edm::Wrapper<l1tp2::CaloPFClusterCollection>" />
```
Comment (aloeliger) [inline DataFormats/L1TCalorimeterPhase2/src/classes_def.xml]: They will want this class version to start from 3.

**Review by aloeliger — CHANGES_REQUESTED** [review]
Please also fix the formatting.

File: DataFormats/L1TCalorimeterPhase2/src/classes_def.xml (hunk 5 above)
Comment (pallabidas) [reply DataFormats/L1TCalorimeterPhase2/src/classes_def.xml]: Changed to 3

File: L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc (hunk 4 above)
Comment (pallabidas) [reply L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc]: Added.

**Review by aloeliger — APPROVED** [review]

**Comment by aloeliger** [comment]
@epalencia I think this can be moved over to the main CMSSW branch to keep it moving, unless you have any further review here. @pallabidas could you open a PR for this material to the main CMSSW fork?

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 1 files that did not meet formatting requirements:
 - L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by epalencia** [comment]
@pallabidas , is this PR in sync with the corresponding one in master, https://github.com/cms-sw/cmssw/pull/42424?

**Comment by pallabidas** [comment]
hi @epalencia these are two different branches, the one merged to master has more edits. Do you want me to apply those changes in this branch?

**Comment by aloeliger** [comment]
@pallabidas If possible, we would like to try to keep this as in-synch with the actual CMSSW as possible.

**Comment by pallabidas** [comment]
@epalencia @aloeliger done, please check.

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 1 files that did not meet formatting requirements:
 - L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by epalencia** [comment]
Tagged as `l1t-phase2-v78`.

**Pull request merged**